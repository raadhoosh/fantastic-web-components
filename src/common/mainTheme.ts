import { Theme } from './styled';
import * as Color from 'color';
function darken(hex: string) {
    return Color(hex)
        .darken(0.5)
        .hex();
}
const mainTheme: Theme = {
    color: {
        primary: '#00aba9',
        primary_dark: darken('#00aba9'),
        secondary: '#0066b1',
        secondary_dark: darken('#0066b1'),
        warning: '#f0ad4e',
        warning_dark: darken('#f0ad4e'),
        light: '#f8f9fa',
        light_dark: darken('#f8f9fa'),
        dark: '#000',
        success: '#d4edda',
        success_dark: '#155724',
        info: '#5bc0de',
        info_dark: darken('#5bc0de'),
        danger: '#d9534f',
        danger_dark: darken('#d9534f'),
    },
    font: {
        size: 16,
        family: '"Roboto"',
    },
    bottom: {
        fontsize: '1rem',
        border: '5px',
    },
    darken: (_color?: string) => {
        return Color(_color)
            .darken(0.03)
            .hex();
    },
    hexToRgba: (hex: string, a: number) => {
        return Color(hex).fade(a ? a : 0.5).string();
    },
};

export default mainTheme;
