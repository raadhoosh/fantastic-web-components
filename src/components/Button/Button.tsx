import * as React from 'react';
import styled from '../../common/styled';

interface IProps {
    /** This is a primary button it out!. */
    primary?: boolean;
    /** This is a secondary button it out!. */
    secondary?: boolean;
    /** This is a success button it out!. */
    success?: boolean;
    /** This is a info button it out!. */
    info?: boolean;
    /** This is a warning button it out!. */
    warning?: boolean;
    /** This is a danger button it out!. */
    danger?: boolean;
    /** This is a light button it out!. */
    light?: boolean;
    /** This is a dark button it out!. */
    dark?: boolean;
    /** This is a text color button it out! */
    children?: any;
    /** This is a Background Color button it out!. */
    backgroundColor?: string;
    /** This is a Text Color button it out!. */
    color?: string;
    /** This is a margin button it out!. */
    margin?: string;
    /** This is a noShadow button it out!. */
    noShadow?: boolean;
    theme?: any;
    size?: 'sm' | 'lg' | 'md';
}

const EmButton = styled('button')(
    (props: (IProps)) => {
        const theme = props.theme.button;
        let shadow = '';
        if (theme.shadow) {
            shadow = theme.shadow;
        }
        let padding = theme.padding;
        let fontsize = theme.fontsize;
        let  lineHeight;
        if (props.size === 'sm') {
            padding = `${props.theme.unit.padding * 8}rem ${props.theme.unit.padding * 7}rem`;
            fontsize = `${props.theme.unit.font * 8}rem`;
            lineHeight = '0.2rem';
        } else if (props.size === 'lg') {
            padding = `${props.theme.unit.padding * 14}rem ${props.theme.unit.padding * 14}rem`;
            fontsize = `${props.theme.unit.font * 11}rem`;
            lineHeight = '0.4rem';
        }
        const btn = {
            'display': 'inline-block',
            'outline': 'none',
            'text-align': 'center',
            'font-weight': 'normal',
            'white-space': 'nowrap',
            'vertical-align': 'middle',
            '-webkit-user-select': 'none',
            '-moz-user-select': 'none',
            '-ms-user-select': 'none',
            'cursor': 'pointer',
            'user-select': 'none',
            'border': theme.border,
            'borderColor': 'transparent',
            'padding': padding,
            'font-size': fontsize,
            'line-height': lineHeight || '1.6rem',
            'border-radius': theme.radius,
            'transition': theme.transition,
            'margin': props.margin ? props.margin : theme.margin,
            'color': props.color ? props.color : theme.color,
            'backgroundColor': 'transparent'
        };
        const color = (props.primary && 'primary') ||
            (props.secondary && 'secondary') ||
            (props.success && 'success') ||
            (props.info && 'info') ||
            (props.danger && 'danger') ||
            (props.warning && 'warning') ||
            (props.light && 'light') ||
            (props.dark && 'dark') ||
            ('');
        if (color !== '') {
            btn.backgroundColor = props.theme.color[color];
            if (!props.noShadow) {
                btn['&:hover'] = {
                    boxShadow: shadow.replace('#color', props.theme.hexToRgba(props.theme.color[color]))
                };
            }
            return btn;
        }

        btn.backgroundColor = props.backgroundColor ? props.backgroundColor : props.theme.color.primary;
        if (!props.noShadow) {
            btn['&:hover'] = {
                boxShadow: shadow.replace('#color', props.theme.hexToRgba(btn.backgroundColor))
            };
        }
        return btn;
    }
);

/**
 * General component description in JSDoc format. Markdown is *supported*.
 */
const Button = (props: IProps) => {

    return (
        <EmButton {...props}> {props.children} </EmButton>
    );
};

export default Button;