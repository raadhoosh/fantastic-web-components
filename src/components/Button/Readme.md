#### Props in default theme
```html
 button: {
        margin: '5px 0px',
        padding: '.375rem .75rem',
        radius: '.85rem',
        fontsize: '1rem',
        color: '#fff',
        border: '5px',
        shadow: '0 0 1rem 0.2rem #color',
        transition: `color .15s ease-in-out,
        background-color.15s ease-in-out,
        border-color .15s ease-in-out,
        box-shadow .25s ease-in-out`
    },
```

### Example:
#### Example Color
```js
<Button primary>Push primary</Button>
<Button secondary>Push secondary</Button>
<Button danger>Push danger</Button>
<Button info>Push info</Button>
<Button success>Push success</Button>
<Button warning>Push warning</Button>
<Button color="#000" light>Push light</Button>
<Button dark>Push dark</Button>
```
#### Example Shadow
```js
<Button noShadow primary>Push primary</Button>
<Button noShadow secondary>Push secondary</Button>
<Button noShadow danger>Push danger</Button>
<Button noShadow info>Push info</Button>
<Button noShadow success>Push success</Button>
<Button noShadow warning>Push warning</Button>
<Button noShadow color="#000" light>Push light</Button>
<Button noShadow dark>Push dark</Button>
```
```js
<br/>
<Button size="lg" warning 
onClick={
    ()=>{
        console.log('clicked me');
    }
}
>click me</Button>
<br/>
```

#### Example Size
```js
<Button size="sm" primary>Push sm</Button>
<Button size="md" primary>Push md</Button>
<Button size="lg" primary>Push lg</Button>

```
