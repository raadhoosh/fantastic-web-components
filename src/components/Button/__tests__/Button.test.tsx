import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { createMatchers } from 'jest-emotion';
import * as emotion from 'emotion';

// Add the custom matchers provided by 'jest-emotion'
expect.extend(createMatchers(emotion));

import Button from '../Button';
import mainTheme from '../../../common/mainTheme';

test('renders with correct styles', () => {
    const btn = renderer.create(<Button theme={mainTheme} danger />).toJSON();
    expect(btn).toHaveStyleRule('background-color', mainTheme.color.danger);
    expect(btn).not.toHaveStyleRule('color', mainTheme.color.danger_dark);
    expect(btn).toMatchSnapshot();
});
