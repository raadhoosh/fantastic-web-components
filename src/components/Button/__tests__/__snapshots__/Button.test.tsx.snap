// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with correct styles 1`] = `
<button
  className="css-13e643k"
>
   
   
</button>
`;
